File: algorithm/dfs/walls_gates.cpp
// https://leetcode.com/problems/walls-and-gates/description/

/*
You are given an m x n grid rooms initialized with these three possible values.

-1 A wall or an obstacle.
0 A gate.
INF Infinity means an empty room. 

We use the value 231 - 1 = 2147483647 to represent INF as you may assume that the distance to a gate is less than 2147483647.
Fill each empty room with the distance to its nearest gate. 
If it is impossible to reach a gate, it should be filled with INF.

Ex1:
Input: rooms = [[2147483647,-1,0,2147483647],[2147483647,2147483647,2147483647,-1],[2147483647,-1,2147483647,-1],[0,-1,2147483647,2147483647]]
Output: [[3,-1,0,1],[2,2,1,-1],[1,-1,2,-1],[0,-1,3,4]]

Ex2:
Input: rooms = [[-1]]
Output: [[-1]]

*/

#include <vector>

using namespace std;

// Time: (m*n), Space: O(m*n)
void wallsAndGates(vector<vector<int>>& rooms) {
    for (int i = 0; i < rooms.size(); ++i) {
        for (int j = 0; j < rooms[i].size(); ++j) {
            if (rooms[i][j] == 0) dfs(rooms, i, j, 0);
        }
    }
}
void dfs(vector<vector<int>>& rooms, int i, int j, int val) {
    if (i < 0 || i >= rooms.size() || j < 0 || j >= rooms[i].size() || rooms[i][j] < val) return;
    rooms[i][j] = val;
    dfs(rooms, i + 1, j, val + 1);
    dfs(rooms, i - 1, j, val + 1);
    dfs(rooms, i, j + 1, val + 1);
    dfs(rooms, i, j - 1, val + 1);
}

File: algorithm/dfs/python/walls_gates.py
"""
Walls and Gates (LeetCode 286)
https://leetcode.com/problems/walls-and-gates/description/

You are given an m x n grid `rooms` with values:
    -1  : wall/obstacle
     0  : gate
    INF : empty room (2**31 - 1)

Fill each empty room with the distance to its nearest gate. If a room cannot
reach any gate, leave it as INF.

Approaches:
    1. DFS from every gate. For each gate cell, run a DFS that walks into
       neighbors only when the neighbor's stored value is strictly greater
       than the new distance we would write. This naturally prunes revisits.
       Time: O(m * n) amortized, Space: O(m * n) recursion in the worst case.
"""

from typing import List


# Time: O(m*n), Space: O(m*n)
def wallsAndGates(rooms: List[List[int]]) -> None:
    for i in range(len(rooms)):
        for j in range(len(rooms[i])):
            if rooms[i][j] == 0:
                dfs(rooms, i, j, 0)


def dfs(rooms: List[List[int]], i: int, j: int, val: int) -> None:
    if (
        i < 0
        or i >= len(rooms)
        or j < 0
        or j >= len(rooms[i])
        or rooms[i][j] < val
    ):
        return
    rooms[i][j] = val
    dfs(rooms, i + 1, j, val + 1)
    dfs(rooms, i - 1, j, val + 1)
    dfs(rooms, i, j + 1, val + 1)
    dfs(rooms, i, j - 1, val + 1)
